lib/std/fmt.myr

```diff
@@ -207,21 +208,33 @@ const fmtval = {sb, ty, ap, params
 	;;
 }
 
+const strreplace = {str, from, to
+        var sparts = std.strsplit(str, from)
+        var sparamstr = std.strjoin(sparts, to)
+        slfree(sparts)
+        -> sparamstr
+}
+
 const parseparams = {paramstr, optdesc
+        const cmask = "_=.-=-_"
 	var params, opts
 	var o, a, ha, gotarg, found
 
 	opts = [][:]
 	if optdesc.len == 0 && paramstr.len > 0
 		std.fatal("invalid format options {}\n")
 	;;
-	params = strsplit(paramstr, ",")
+        /* Mask '\\,', so that we can use it in case we need a ',' in a format string, as in: */
+        /* 'std.put("{f=year is %Y\\, the day is %d}\n", date.mkdate(2017, 10, 30, ""))'. */
+        var sparamstr = strreplace(paramstr, "\\,", cmask)
+	params = strsplit(sparamstr, ",")
 	for p : params
```

Comment:
Hm. The problem being solved is a good one to fix, but the solution here
feels a bit hacky.

Probably the cleanest ting e turn getparams/parseparams into a small state
machine,similar to how we handle `{` and `}` in sbfmtv.  Even better may be
to merge all of the format string parsing. That gets us to  something like:

	const sbfmtv = {...
		...
		while fmt.len != 0
			(c, fmt) = strstep(fmt)
			match (state, c)
			| (`Outer, '{'):
				if decode(fmt) == '{'
					(c, fmt) = strste(fmt)
					sbputc(sb, '{')
				else
					state = `Params
				;;
			| (`Params, '\\'): ...
			| (`Params, ','): ...
			| (`Params, '}': ...
			;;
		;;
	}

Doing this should also make it easy to implement escaping the closing '}' in a
format string, which we might as well fix here too.

Not sure if we should change how we escape `{` for consistency.

Reply:
Yes, you're right - I'm not too proud about that hack; still it just fits my very limited Myrddin foo ;) I'll need to put aside some more time to try to understand your proposal and work on it.
Thanks, Frank

Reply:
No worries, I can give a bit more detail, but the idea is just to write a finite state machine (https://patterns.eecs.berkeley.edu/?page_id=470) that handles these things in a single pass through the string. In order to escape, when you see a '\\c' while inside a param list, that just decodes to the next character in the param list.